primitives/tests/light_client/mod.rs

```diff
@@ -0,0 +1,231 @@
+		let validator_set_root = commitment
+			.payload
+			.next_validator_set
+			.as_ref()
+			.ok_or(Error::NoValidatorSetInPayload)?;
+		if !validator_set_proof.is_valid(validator_set_root) {
+			return Err(Error::InvalidValidatorSetProof);
+		}
+		let set = validator_set_proof.into_data();
+
+		let new_id = self.validator_set.0 + 1;
+		self.validator_set = (new_id, set);
+		self.last_commitment = Some(commitment);
+
+		Ok(())
+	}
+
+	pub fn verify_proof<R>(&self, proof: merkle_tree::Proof<Mmr, R>) -> Result<R, Error> {
+		if proof.is_valid(&self.last_payload().mmr) {
+			Ok(proof.into_data())
+		} else {
+			Err(Error::InvalidMmrProof)
+		}
+	}
+
+	pub fn validator_set(&self) -> &(ValidatorSetId, Vec<validator_set::Public>) {
+		&self.validator_set
+	}
+
+	pub fn last_commitment(&self) -> Option<&Commitment> {
+		self.last_commitment.as_ref()
+	}
+
+	pub fn last_payload(&self) -> &Payload {
+		&self
+			.last_commitment()
+			.expect("Genesis doesn't contain commitment.")
+			.payload
+	}
+
+	fn validate_commitment(&self, commitment: SignedCommitment) -> Result<Commitment, Error> {
+		let no_of_non_empty_signatures = commitment.no_of_signatures();
+		let SignedCommitment { commitment, signatures } = commitment;
+		// Make sure it's signed by the current validator set we know of.
+		if self.validator_set.0 != commitment.validator_set_id {
+			return Err(Error::InvalidValidatorSetId {
+				expected: self.validator_set.0,
+				got: commitment.validator_set_id,
+			});
+		}
+
+		// Make sure it's not worse than what we already have.
+		let best_block = self.last_commitment().map(|c| c.block_number).unwrap_or(0);
+		if commitment.block_number <= best_block {
+			return Err(Error::OldBlock {
+				best_known: best_block,
+				got: commitment.block_number,
+			});
+		}
+
+		// check number of signatures
+		let validator_set_count = self.validator_set.1.len();
+		if signatures.len() != validator_set_count {
+			return Err(Error::InvalidNumberOfSignatures {
+				expected: validator_set_count,
+				got: signatures.len(),
+			});
+		}
+
+		// check the validity of signatures
+		let minimal_number_of_signatures = self.minimal_number_of_signatures();
+		if no_of_non_empty_signatures < minimal_number_of_signatures {
+			return Err(Error::NotEnoughValidSignatures {
+				expected: minimal_number_of_signatures,
+				got: no_of_non_empty_signatures,
+				valid: None,
+			});
+		}
+
+		let mut valid = 0;
+		for (signature, public) in signatures.into_iter().zip(self.validator_set.1.iter()) {
+			match signature {
+				Some(signature) if signature.is_valid_for(&public) => {
+					valid += 1;
+				}
+				_ => {}
+			}
+		}
+
+		if valid < minimal_number_of_signatures {
+			return Err(Error::NotEnoughValidSignatures {
+				expected: minimal_number_of_signatures,
+				got: no_of_non_empty_signatures,
+				valid: Some(valid),
+			});
+		}
+
+		Ok(commitment)
+	}
+
+	fn minimal_number_of_signatures(&self) -> usize {
+		(2 * self.validator_set.1.len() + 2) / 3
```

Comment:
```suggestion
		(2 * self.validator_set.1.len() + 1) / 3
```

Reply:
Hmm, for `2` validators we still need both of them to sign, while `(2*2 + 1) / 3 = 1`, similarly for `5` we want at least 4, so `(2 * 5 + 2) / 3 = 4`

So I think this is correct.

Reply:
For N = 3 this method will return 2, whereas we need the 3 of them. I think the main issue (that I didn't notice earlier) is that we are adding the `+ 2` before doing the division. `2 * self.validator_set.1.len() / 3 + 1` should do the right thing.

Reply:
You're right. I thought we need `ceil(2/3 * X)`, but we actually need `floor(2/3 * X + 1)`